simple db based target locker

tfg13 opened the pull request:
Summary of changes:
 - Add a minimal rdbms-based target locker implementation
 - Clarify target.Locker API docs


This adds a minimal target locker that uses a relational database as backend.

Using this locker allows sharing locks between multiple instances of Contest Server and makes locks survive server restarts.

The unit/integration test suite for this new locker is an extended version of the inmemory test suite, some tests changed to confirm with the clarified locker API and some tests added to prevent bugs like #134.

This implementation only uses target.ID as primary and only identifier for targets, we have a refactor pending to make it more clear how targets are specified and what is required to identify them.

insomniacslk commented on plugins/targetlocker/dblocker/dblocker.go:
```diff
@@ -0,0 +1,310 @@
+// cleanExpired deletes all expired locks on the given targets
+// (lock owner is ignored)
+func (d *DBLocker) cleanExpired(tx *sql.Tx, targets []string, now time.Time) error {
+	q := "DELETE FROM locks WHERE expires_at < ? AND target_id IN " + listQueryString(len(targets)) + ";"
```
yeah but at least this can be modified by the caller (e.g. for integ testing), which is more future-proof

tfg13 replied:
I am going to punt this one to a future PR. It opens the "concat query strings" can of worms again, and at that point we probably should use something like safesql, but that is too big a refactor to integrate it now.

insomniacslk commented on plugins/targetlocker/dblocker/dblocker.go:
```diff
@@ -0,0 +1,310 @@
+
+// cleanExpired deletes all expired locks on the given targets
+// (lock owner is ignored)
+func (d *DBLocker) cleanExpired(tx *sql.Tx, targets []string, now time.Time) error {
```
`len(targets)` should be checked to be > 0

insomniacslk replied:
this is marked as resolved - but why don't you check for zero length, instead of executing the DB query anyway?
```
if len(targets) == 0 {
    return nil
}
```

tfg13 replied:
I do, but right away in `Unlock`, `Lock`, and `RefreshLocks`, so we won't get here without locks :)

Reason I put it there is that there are 2 other places where it can create invalid queries with `... IN ()` (basically everywhere `listQueryString` is used)

insomniacslk commented on plugins/targetlocker/dblocker/dblocker.go:
```diff
@@ -0,0 +1,310 @@
+		return "()"
+	case 1:
+		return "(?)"
+	default:
```
also please add a comment on why this is safe from a SQL security perspective

insomniacslk commented on plugins/targetlocker/dblocker/dblocker.go:
```diff
@@ -0,0 +1,310 @@
+
+// DBLocker implements a simple target locker based on a relational database.
+// The current implementation only supports MySQL officially.
+// All functions in DBLocker are safe for concurrent by multiple goroutines.
```
typo? "for concurrent use" I assume

insomniacslk commented on plugins/targetlocker/dblocker/dblocker.go:
```diff
@@ -0,0 +1,310 @@
+	if err != nil {
+		return nil, fmt.Errorf("unable to read existing locks: %w", err)
+	}
+	defer rows.Close()
```
*cough* (just kidding, the docs say that `Close` is idempotent, and users should check `sql.Rows.Err()` instead, as documented at https://golang.org/pkg/database/sql/#Rows )

insomniacslk commented on plugins/targetlocker/dblocker/dblocker.go:
```diff
@@ -0,0 +1,310 @@
+			return nil, fmt.Errorf("unexpected read from database: %w", err)
+		}
+		locks[row.targetId] = row
+	}
```
(as noted above, `rows.Err()` should be checked here)

tfg13 replied:
Oh good point!

insomniacslk commented on plugins/targetlocker/dblocker/dblocker.go:
```diff
@@ -0,0 +1,310 @@
+	// locking is all or nothing in one transaction
+	// phantom reads might mess with the expired lock cleaning assumptions below,
+	// so request serializable isolation
+	tx, err := d.db.BeginTx(context.Background(), &sql.TxOptions{Isolation: sql.LevelSerializable})
```
I'm no SQL expert - does `sql.LevelSerializable` also guarantee the desired consistency level? I.e. that  lock and subsequent read will always be consistent on any primary/replica?
At a quick look isolation and consistency should not be confused, and I can't confirm that this is what we want for DB locking

tfg13 replied:
Not sure I am following here. I am using this purely for query isolation, since some steps assume previous steps are done properly. Example: When locking I read other locks and check them (because we need this to produce a useful error anyway). The DB will enforce only one lock per asset on its own of course, but it makes the code simple to assume there won't be any conflicts after I run the check.

What are you asking about exactly wrt. to consistency, do you mean strict read-after-write consistency for multiple queries to the same DB?

insomniacslk replied:
we discussed this internally already, and concluded that transactions paired to serializable isolation should be good enough for locking purposes. To answer to your question, yes, I meant strict read-after-write consistency

insomniacslk commented on plugins/targetlocker/dblocker/dblocker.go:
```diff
@@ -0,0 +1,310 @@
+		return err
+	}
+	// go through existing locks, they are either held by something else (abort)
+	// help by us (update time), or not held (insert)
```
"help" -> "held"

insomniacslk commented on plugins/targetlocker/dblocker/dblocker.go:
```diff
@@ -0,0 +1,310 @@
+		}
+	}
+	if len(conflicts) > 0 {
+		return fmt.Errorf("unable to lock, targets have conflicting locks: %v", conflicts)
```
we probably also want the job ID and the target ID in this error string, for sake of debugging when necessary

tfg13 replied:
They are! This is why there is a pretty printer on the dblock struct :)

insomniacslk replied:
I mean not just the info on what is currently holding the lock, but also on what tried to get the lock but failed

tfg13 replied:
done, looks like this now:
```
unable to lock targets [001 002] for owner 124, have conflicting locks: [target: 001 job: 123 expires: 2020-08-25 15:01:48 +0000 UTC target: 002 job: 123 expires: 2020-08-25 15:01:48 +0000 UTC]
```

insomniacslk reviewed (commented):
I have reviewed up until `handleUnlock` (excluded). Publishing this second part, then continuing

insomniacslk commented on plugins/targetlocker/dblocker/dblocker.go:
```diff
@@ -0,0 +1,310 @@
+		}
+	}
+	if len(conflicts) > 0 {
+		log.Warningf("unable to unlock some targets due to different lock owners: %+v", conflicts)
```
let's add the job ID that is failing the request here too

tfg13 replied:
same as above

insomniacslk replied:
same as above :P

tfg13 replied:
done, same as above :P

insomniacslk commented on plugins/targetlocker/dblocker/dblocker.go:
```diff
@@ -0,0 +1,310 @@
+
+// queryLocks returns a map of ID -> dblock for a given list of targets
+func (d *DBLocker) queryLocks(tx *sql.Tx, targets []string) (map[string]dblock, error) {
+	q := "SELECT target_id, job_id, expires_at FROM locks WHERE target_id IN " + listQueryString(len(targets))
```
the trailing ";" is missing here and present elsewhere, and we should probably be consistent

tfg13 replied:
I just checked the official examples - it is also inconsistent there. AAAAA
Fixing :)

insomniacslk replied:
yeah :( it is safe with mysql and probably other engines too, but I guess it doesn't hurt to avoid bad karma :)

insomniacslk commented on plugins/targetlocker/dblocker/dblocker.go:
```diff
@@ -0,0 +1,310 @@
+	}
+	_, err = tx.Exec(del, queryList...)
+	if err != nil {
+		return fmt.Errorf("unable to unlock targets: %w", err)
```
and job ID here too

insomniacslk commented on plugins/targetlocker/dblocker/dblocker.go:
```diff
@@ -0,0 +1,310 @@
+// ResetAllLocks resets the database and clears all locks, regardless of who owns them.
+// This is primarily for testing, and should not be used by used in prod, this
+// is why it is not exposed by target.Locker
+func (d *DBLocker) ResetAllLocks() error {
```
being internal, unexport it?

tfg13 replied:
No then the semi-external integration tests can't reach it :/

tfg13 commented on plugins/targetlocker/dblocker/dblocker.go:
```diff
@@ -0,0 +1,310 @@
+func targetIDList(targets []*target.Target) []string {
+	res := make([]string, 0, len(targets))
+	for _, target := range targets {
+		res = append(res, target.ID)
```
ok I am bringing the verify function back

tfg13 replied:
added `validateTargets` to make sure none are empty

tfg13 replied:
also added a unit test

tfg13 commented:
Ok addressed all comments except the one about making the table name a parameter, punting this one as we probably should use safesql then, which is a bigger rework.

insomniacslk commented on plugins/targetlocker/dblocker/dblocker.go:
```diff
@@ -0,0 +1,310 @@
+}
+
+// String pretty-prints dblocks for logging and errors
+func (d dblock) String() string {
```
I prefer the latter :P (also, that's `%+v`, not `%#+v`) but it's personal, I'm OK with the current one
